File: solutions/fsharp/meetup/1/Meetup.fs
module Meetup

open System

type Week = First | Second | Third | Fourth | Last | Teenth

type DayOfWeek = 
    | Sunday = 0    
    | Monday = 1
    | Tuesday = 2
    | Wednesday = 3
    | Thursday = 4
    | Friday = 5
    | Saturday = 6


let meetup year month week (dayOfWeek: DayOfWeek): DateTime =
    let days =    
        match week with
        | First -> [1..7]
        | Second -> [8..14]
        | Third -> [15..21]
        | Fourth -> [22..28]
        | Last ->
            match month with
            | 2 ->
                match year % 4 with
                | 0 -> [23..29]
                | _ -> [22..28]
            | 4 | 6 | 9 | 11 -> [24..30]
            | _ -> [25..31]
        | Teenth -> [13..19]
    let day = 
        days
        |> List.filter (fun day -> DateTime(year, month, day).DayOfWeek |> int = int dayOfWeek)
        |> List.exactlyOne
    
    DateTime(year, month, day)

File: solutions/erlang/meetup/1/src/meetup.erl
-module(meetup).
-export([meetup/4]).

dow(DayOfWeek) ->
  case DayOfWeek of
    monday    -> 1;
    tuesday   -> 2;
    wednesday -> 3;
    thursday  -> 4;
    friday    -> 5;
    saturday  -> 6;
    sunday    -> 7
  end.

first(Year, Month, DayOfWeek) -> 1 + (7 + dow(DayOfWeek) - calendar:day_of_the_week(Year,Month,1)) rem 7.

meetup(Year, Month, DayOfWeek, Week) -> 
  case Week of
    first  -> {Year, Month, first(Year, Month, DayOfWeek)}; 
    second -> {Year, Month, first(Year, Month, DayOfWeek) + 7};
    third  -> {Year, Month, first(Year, Month, DayOfWeek) + 2 * 7}; 
    fourth -> {Year, Month, first(Year, Month, DayOfWeek) + 3 * 7};
    teenth ->
      case first(Year, Month, DayOfWeek) + 2*7 < 20 of
        true  -> {Year, Month, first(Year, Month, DayOfWeek) + 2 * 7};
        false -> {Year, Month, first(Year, Month, DayOfWeek) + 7 }
      end;
    last -> 
      case Month of
        12 -> calendar:gregorian_days_to_date(calendar:date_to_gregorian_days({Year+1, 1, first(Year+1, 1, DayOfWeek)}) - 7 );
        _  -> calendar:gregorian_days_to_date(calendar:date_to_gregorian_days({Year, Month + 1, first(Year, Month+1, DayOfWeek)}) - 7)
      end
  end.
